include/metal/lambda/arg.hpp

```diff
@@ -1,24 +1,63 @@
 // Copyright Bruno Dutra 2015-2016
 // Distributed under the Boost Software License, Version 1.0.
-// (See accompanying file LICENSE.txt or copy at http://boost.org/LICENSE_1_0.txt)
+// See accompanying file LICENSE.txt or copy at http://boost.org/LICENSE_1_0.txt
 
 #ifndef METAL_LAMBDA_ARG_HPP
 #define METAL_LAMBDA_ARG_HPP
 
+#include <metal/list/list.hpp>
+#include <metal/list/at.hpp>
+#include <metal/lambda/lambda.hpp>
+#include <metal/number/number.hpp>
+
 #include <cstddef>
 
 namespace metal
 {
     namespace detail
     {
         template<std::size_t n>
-        struct arg;
+        struct arg_impl
+        {
+            template<typename... vals>
+            using impl = at<list<vals...>, size_t<n - 1>>;
+
+            using type = lambda<impl>;
+        };
+
+        template<>
+        struct arg_impl<2U>
+        {
+            template<typename, typename val, typename...>
+            using impl = val;
+
+            using type = lambda<impl>;
+        };
+
+        template<>
+        struct arg_impl<1U>
+        {
+            template<typename val, typename...>
+            struct lazy
+            {
+                using type = val;
+            };
+
+            template<typename... args>
+            using impl = typename lazy<args...>::type;
```

Comment:
maybe we can work around that [anoying VC bug](https://github.com/brunocodutra/metal/commit/a80937df23e45b9c780f11b0c8117cab29796587) here by changing `impl` to

    template<typename val, typename = void, typename...>
    using impl = val;

instead delegating to `lazy`.